src/Stratis.Bitcoin/AddressIndexing/AddressIndexer.cs

```diff
@@ -0,0 +1,284 @@
+﻿using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using LiteDB;
+using Microsoft.Extensions.Logging;
+using NBitcoin;
+using Stratis.Bitcoin.Configuration;
+using Stratis.Bitcoin.EventBus;
+using Stratis.Bitcoin.EventBus.CoreEvents;
+using Stratis.Bitcoin.Interfaces;
+using Stratis.Bitcoin.Primitives;
+using Stratis.Bitcoin.Signals;
+using Stratis.Bitcoin.Utilities;
+using Script = NBitcoin.Script;
+
+namespace Stratis.Bitcoin.AddressIndexing
+{
+    public class AddressIndexer : IDisposable
+    {
+        private readonly ISignals signals;
+
+        private readonly NodeSettings nodeSettings;
+
+        private readonly Network network;
+
+        private readonly IBlockStore blockStore;
+
+        private readonly INodeStats nodeStats;
+
+        private readonly ILogger logger;
+
+        private readonly DataFolder dataFolder;
+
+        private const string DbKey = "AddrData";
+
+        private SubscriptionToken blockConnectedSubscription, blockDisconnectedSubscription;
+
+        /// <summary>Queue that is populated when block is connected or disconnected.</summary>
+        /// <remarks><c>bool</c> key is <c>true</c> when block is connected.</remarks>
+        private AsyncQueue<KeyValuePair<bool, ChainedHeaderBlock>> blockReceivedQueue;
+
+        private LiteDatabase db;
+
+        private LiteCollection<AddressIndexerData> dataStore;
+
+        private AddressIndexerData data;
+
+        public AddressIndexer(NodeSettings nodeSettings, ISignals signals, DataFolder dataFolder, ILoggerFactory loggerFactory,
+            Network network, IBlockStore blockStore, INodeStats nodeStats)
+        {
+            this.signals = signals;
+            this.nodeSettings = nodeSettings;
+            this.network = network;
+            this.blockStore = blockStore;
+            this.nodeStats = nodeStats;
+            this.dataFolder = dataFolder;
+
+            this.logger = loggerFactory.CreateLogger(this.GetType().FullName);
+        }
+
+        public void Initialize()
+        {
+            if (this.nodeSettings.TxIndex)
+            {
+                string dbPath = Path.Combine(this.dataFolder.RootPath, "addressindex.litedb");
+                this.db = new LiteDatabase(new ConnectionString() {Filename = dbPath});
+
+                this.logger.LogDebug("TxIndexing is enabled.");
+
+                this.dataStore = this.db.GetCollection<AddressIndexerData>(DbKey);
+
+                this.data = this.dataStore.FindAll().FirstOrDefault();
```

Comment:
`this.data` can surely have a better name (same goes for `this.dataStore` its not clear what data store is we don't need a comment just a good variable name)

Reply:
fixed